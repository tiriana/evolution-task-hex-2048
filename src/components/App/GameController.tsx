import React from "react";

import { useState } from "react";
import { GameConfig } from "./GameConfig";
import SimpleBoard from "./SimpleBoard";
import { MoveListener } from "./MoveListener";
import { GameStateBadge } from "./GameStateBadge";

import Direction from "../../logic/Direction";
import Cell from "../../logic/Cell";

import { default as BoardLogic } from "../../logic/Board";
import { GetData } from "./GetData";
import { RngServerResponse } from "./RngServerResponse";
import { HexSimple } from "./HexSimple";
import { RngServerRequest } from "./RngServerRequest";

import env from "./env";

const cellToSimpleHex: (cell: Cell) => HexSimple = (cell: Cell) => {
  return {
    x: cell.x,
    y: cell.y,
    z: cell.z,
    value: 0 | cell.value,
  };
};

const boardToRequest: (board: BoardLogic) => RngServerRequest = (
  board: BoardLogic
) => {
  const request: RngServerRequest = [];

  board.cells.forEach((cell: Cell) => {
    cell.value && request.push(cellToSimpleHex(cell));
  });

  return request;
};

declare global {
  interface Window {
    bag: any;
  }
}

const GameController: React.FC<GameConfig> = ({ hostname, port, radius }) => {
  const [waitingForData, setWaitingForData] = useState(true);
  const [waitingForInput, setWaitingForInput] = useState(false);
  const [response, setResponse] = useState([] as RngServerResponse);

  const [board, setBoard] = useState(new BoardLogic(radius - 1));

  window.bag = {
    board,
    Direction,
  };

  window.bag = {
    board,
    Direction,
  };

  const onMove = (dir: Direction) => {
    const before = board.values().join();
    board.makeMove(dir);
    const after = board.values().join();

    if (before !== after) {
      setBoard(board);
      setWaitingForInput(false);
      setWaitingForData(true);
    }
  };

  const onReceived = (response: RngServerResponse) => {
    setWaitingForData(false);

    response.forEach(({ x, y, z, value }: HexSimple) => {
      board.getCell(x, y, z).value = value;
    });

    setResponse(response);

    setWaitingForInput(true);
  };

  const debug = (
    <pre
      style={{
        position: "fixed",
        right: 0,
      }}
    >
      {JSON.stringify(
        {
          hostname,
          port,
          radius,
          response,
          waitingForData,
          waitingForInput,
        },
        null,
        2
      )}
    </pre>
  );

  return (
    <div className={"mainContainer"}>
      {waitingForData && (
        <GetData
          {...{ hostname, port, radius }}
          onReceived={onReceived}
          request={boardToRequest(board)}
        />
      )}
      {waitingForInput && board.hasMove() && <MoveListener onMove={onMove} />}

      {env.debug && debug}

      <SimpleBoard board={board} />
      <GameStateBadge hasMove={board.hasMove()} />
    </div>
  );
};

export default GameController;
