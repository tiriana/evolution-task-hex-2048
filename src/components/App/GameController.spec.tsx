import { render } from "@testing-library/react"; // cleanup is automatic
import GameController from "./GameController";
import axios from "axios";
import { act } from "react-dom/test-utils";
jest.mock("axios");
const tick = async () => new Promise((r) => setImmediate(r));

it("renders the game", async () => {
  const [x, y, z, value] = [0, 1, -1, 128];
  const response = {
    data: [{ x, y, z, value }],
  };
  const responsePromise = new Promise((res) =>
    setImmediate(() => res(response), 1)
  );
  axios.post.mockReturnValueOnce(responsePromise);

  const component = render(
    <GameController hostname="hostname" port={13337} radius={2} />
  );

  expect(component.container.innerHTML).toMatchSnapshot();

  await act(tick);

  expect(
    component.container.querySelector(
      `[data-x="${x}"][data-y="${y}"][data-z="${z}"]`
    )
  ).toHaveAttribute("data-value", "" + value);

  expect(component.container.innerHTML).toMatchSnapshot();
});
