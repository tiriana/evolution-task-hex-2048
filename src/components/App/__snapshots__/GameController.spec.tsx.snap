// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the game 1`] = `"<div class=\\"mainContainer\\"><div class=\\"board\\" style=\\"transform: scale(1);\\"><div><div data-value=\\"0\\" data-x=\\"-1\\" data-y=\\"0\\" data-z=\\"1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"-1\\" data-y=\\"1\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"0\\" data-y=\\"-1\\" data-z=\\"1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"0\\" data-y=\\"0\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"0\\" data-y=\\"1\\" data-z=\\"-1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"1\\" data-y=\\"-1\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"1\\" data-y=\\"0\\" data-z=\\"-1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div></div></div><div data-status=\\"playing\\">Game status: playing</div></div>"`;

exports[`renders the game 2`] = `"<div class=\\"mainContainer\\"><div class=\\"board\\" style=\\"transform: scale(1);\\"><div><div data-value=\\"0\\" data-x=\\"-1\\" data-y=\\"0\\" data-z=\\"1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"-1\\" data-y=\\"1\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"0\\" data-y=\\"-1\\" data-z=\\"1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"0\\" data-y=\\"0\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"128\\" data-x=\\"0\\" data-y=\\"1\\" data-z=\\"-1\\" class=\\"cell\\"><div class=\\"inner value--128\\">128</div></div><div data-value=\\"0\\" data-x=\\"1\\" data-y=\\"-1\\" data-z=\\"0\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div><div data-value=\\"0\\" data-x=\\"1\\" data-y=\\"0\\" data-z=\\"-1\\" class=\\"cell\\"><div class=\\"inner value--0\\"></div></div></div></div><div data-status=\\"playing\\">Game status: playing</div></div>"`;
